solutions/usaco-623.mdx

```diff
@@ -2,115 +2,140 @@
+
+	int size(int x) { return -e[get(x)]; }
+
+	bool unite(int x, int y) {	// union by size
+		x = get(x), y = get(y);
+		if (x == y) return false;
+		if (e[x] > e[y]) swap(x, y);
+		e[x] += e[y];
+		e[y] = x;
+		return true;
 	}
-	top[b.ff][b.ss] = a;
-	size[a.ff][a.ss] += size[b.ff][b.ss];
-	return true;
-}
+};
 //EndCodeSnip
 
 int main() {
-	ifstream fin("fencedin.in");
+	freopen("fencedin.in", "r", stdin);
+	freopen("fencedin.out", "w", stdout);
+
+	int hdist, vdist, n, m;
+	cin >> hdist >> vdist >> n >> m;
+
+	vector<int> vx(n + 1), vy(m + 1);
 
-	// Sorting the fence lengths from shortest to longest
-	int xlimit, ylimit, n, m;
-	fin >> xlimit >> ylimit >> n >> m;
-	vector<int> vert(n + 1);
-	vector<int> horz(m + 1);
 	for (int i = 1; i <= n; i++) {
-		fin >> vert[i];
+		cin >> vx[i];
 	}
+
 	for (int i = 1; i <= m; i++) {
-		fin >> horz[i];
+		cin >> vy[i];
 	}
-	sort(vert.begin(), vert.end());
-	sort(horz.begin(), horz.end());
-	for (int i = 0; i < n; i++) {
-		vert[i] = vert[i + 1] - vert[i];
-	}
-	for (int i = 0; i < m; i++) {
-		horz[i] = horz[i + 1] - horz[i];
-	}
-	vert[n] = xlimit - vert[n];
-	horz[m] = ylimit - horz[m];
-	sort(vert.begin(), vert.end());
-	sort(horz.begin(), horz.end());
-
-	// Set up the DSU
-	vector<vector<int>> size(n + 1, vector<int>(m + 1, 1));
-	vector<vector<pii>> top(n + 1, vector<pii>(m + 1, make_pair(0, 0)));
-	for (int i = 0; i <= n; i++) {
-		for (int j = 0; j <= m; j++) {
-			top[i][j] = make_pair(i, j);
+
+	vx.push_back(hdist);
+	vy.push_back(vdist);
+
+	sort(vx.begin(), vx.end());
+	sort(vy.begin(), vy.end());
+
+	// add fences at borders to calculate edges at front/back
+	n += 2;
+	m += 2;
+
+	/*
+	 * we'll number sections from left
+	 * to right.
+	 *
+	 * -------------
+	 * |  1  |  2  |
+	 * |-----------|
+	 * |  3  |  4  |
+	 * -------------
+	 */
+	vector<Edge> edges;
+
+	// vertical fence edges
+	int sct = 0, e = 1;
+	while (e < m) {
+		// all vertical fences in one row have the same length
+		for (int i = 0; i < n - 2; i++) {
+			edges.push_back(Edge{vy[e] - vy[e - 1], sct, sct + 1});
+			sct++;
 		}
+
+		sct++;
+		e++;
 	}
 
-	// Perform the MST
-	ll ans = 0;
-	for (int x = 0, y = 0; x <= n || y <= m; ) {
-		if (x <= n && vert[x] < horz[y] || y > m) {
-			for (int i = 0; i < m; i++) {
-				if (merge(make_pair(x, i), make_pair(x, i + 1), top, size)) {
-					ans += vert[x];
-				}
-			}
-			x++;
-		} else {
-			for (int i = 0; i < n; i++) {
-				if (merge(make_pair(i, y), make_pair(i + 1, y), top, size)) {
-					ans += horz[y];
-				}
-			}
-			y++;
+	// horizontal fence edges
+	sct = n - 1, e = 1;
+	while (e < n) {
+		int init = sct;
+		// all horizontal fences in one column have the same length
+		for (int i = 0; i < m - 2; i++) {
+			edges.push_back(Edge{vx[e] - vx[e - 1], sct - (n - 1), sct});
+
+			// go down a row
+			sct += (n - 1);
 		}
+		sct = init + 1;
+		e++;
 	}
 
-	ofstream fout("fencedin.out") << ans << endl;
+	// naive kruskal's implementation
```

Comment:
wdym naive lmao

Reply:
oops

Reply:
[was typing my english essay while doing this (meant to type naive in the google doc)] 💀